### PR TITLE
Adds steel mill location

### PR DESCRIPTION
**Pull request by Kanexan**
#### Summary
```SUMMARY: Content "Adds steel mill location to the game."```

#### Purpose of change
To add another major industrial site, representing a major New England industry, and making a place for finding amounts of steel, iron, and aluminum similar to how the Lumbermill provides large amounts of wood planks and panels.

#### Describe the solution
This adds the Steel Mill and accompanying small cargo rail depot to the game, although it does not have item, vehicle, or monster spawns as of yet. It's a large, multi-story and multi-building complex, with a distinct "office" section, the main mill area, a storage building, the railway itself with a control building and two large rail cranes for unloading and reloading shipping containers, and a smaller building serving as an on-site foundry for high-precision or custom order steel and aluminum parts. 

The office section will have more normal office clutter, while the main mill should have iron ore, and steel ingots, sheets, and rebar spawn as items. The foundry will have steel and aluminum ingots, as well as miscellaneous other items made of the same two materials. The mill as a whole will also be a prime place to find smelting and forging materials, such as welding goggles, metal tongs, welders, crucibles, etc.

In the future, I want to make the roof area more detailed, add the monster, vehicle, and item spawn lists, and possibly create some additional heavy-industrial zombie types for a more detailed and realistic feel to the area.

#### Describe alternatives you've considered
Creating a smaller, more compact steel foundry site to the game, as opposed to the large industrial complex that is a modern steel mill.

####Screenshots

Screenshot (28)
Screenshot (26)
The two floors of the attached offices and lobby.

Screenshot (22)
Screenshot (27)
The furnace section of the main mill building, with some of the additional machinery and the foreman's office included.

Screenshot (23)
The foundry building; this is where custom-ordered and precision-quality parts will be made and found, as opposed to the general steel-production of the main mill.

Screenshot (24)
Screenshot (25)
The two z-levels of the accompanying cargo rail depot. The pillars are part of the (technically) movable railway cranes, but seeing as functionally there is no possible way the player could ever actually move either of them after the Cataclysm, I'm representing them as static objects here.

### CODE REVIEW

**Comment by curstwist** [comment]
looks good, I'll be happy to check this out in game later today to check your terrains under furniture. 
I don't see fill_ter entries in mapgen which would cut down on the amount of furniture you'll need to specify terrains for (depends on if you use the same floor on each level/building).

A few notes: 
Maybe add gutters to the roof.  

I'm also not seeing any items in your maps.  I see you are planning on adding monsters and vehicles in a 2nd PR?  

Could drop some basic zombies in for now so it's not a massive free base or set occurrences to [ 0, 0 ] until it's complete with spawns.

Over all, looks great & can't wait to see it finished.

**Comment by Kanexan** [comment]
Mm. That terrains under furniture is going to be an issue for the second z-level; I'll need to look into how to fix that. And yeah, I'm planning on either doing an immediate second PR for the item/zombie/vehicle spawns, or adding them into this one.
And thanks for the feedback! I'm glad you think it looks good!

**Comment by curstwist** [comment]
If you make the fill_ter the dominate floor terrain type per level.  Any furniture that doesn't have their symbol also defined as a floor type in terrain will use the fill_ter instead.  Then you only need to add the extra terrain entries for rarer floor types. (if this doesn't make sense, catch me on discord).

File: data/json/mapgen/steel_mill/steel_mill_z2.json [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        ":::::]MMMMGaaaaaaGGGGGGGGGGGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FFFFGaaaaaaGGGGGGGGGGGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FaaFGaaaaaaaaaaaaaaaaGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FaaFGaaaaaaaaaaaaaaaaGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FFFFGaaaaaaaaaaaaaaaaGGRrrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]MMMMGaaaaaaaaaaaaaaaaGGRrrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FFFFGaaaaaaaaaaaaaaGGGG]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
+        ":::::]FaaFGaaaaaaaaaaaaaaGGGG]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
+        ":::::]FaaFGGaaaaaaaaaaaaaGGG<]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
+        ":::::]FFFFEGaaaaaaaaaaaaaGGG<]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
```
Comment (anothersimulacrum) [inline data/json/mapgen/steel_mill/steel_mill_z2.json]:
```suggestion
        ":::::]FFFFEGaaaaaaaaaaaaaGGG>]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
```

File: data/json/mapgen/steel_mill/steel_mill_z2.json [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        ":::::]FFFFGaaaaaaGG]]]]]]]]]]]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
+        ":::::]MMMMGaaaaaaGGGGGGGGGGGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FFFFGaaaaaaGGGGGGGGGGGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FaaFGaaaaaaaaaaaaaaaaGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FaaFGaaaaaaaaaaaaaaaaGG]rrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FFFFGaaaaaaaaaaaaaaaaGGRrrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]MMMMGaaaaaaaaaaaaaaaaGGRrrrrrrrrrrrrrrrrr:::::::::::::rrrrrrrrrr::",
+        ":::::]FFFFGaaaaaaaaaaaaaaGGGG]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
+        ":::::]FaaFGaaaaaaaaaaaaaaGGGG]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
+        ":::::]FaaFGGaaaaaaaaaaaaaGGG<]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
```
Comment (anothersimulacrum) [inline data/json/mapgen/steel_mill/steel_mill_z2.json]:
```suggestion
        ":::::]FaaFGGaaaaaaaaaaaaaGGG>]rrrrrrrrrrrrrrrrr:::::::::::::::::::::::::",
```

File: data/json/overmap/overmap_special/specials.json [hunk 3]
```diff
@@ -5511,5 +5511,71 @@
     "city_sizes": [ 3, 20 ],
     "occurrences": [ 0, 0 ],
     "flags": [ "UNIQUE" ]
+  },
+  {
+    "type": "overmap_special",
+    "id": "steel mill",
+    "overmaps": [
+      { "point": [ 0, 0, 0 ], "overmap": "steel_mill_0_1" },
```
Comment (anothersimulacrum) [inline data/json/overmap/overmap_special/specials.json]: These all a direction suffix (`_north`).

File: data/json/mapgen/steel_mill/steel_mill_z1.json [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        "v....]P  P                ee@]++]]]]]]]]]]] b ,____________............v",
+        "v....]P  P  <<  mc m      eee]   hTh hTh f]b   ____________''''''''''''v",
+        "v....]MMMMPPPPPPPME       eee]           C]  b,____________''OOOOOOO'''v",
+        "v....]MEEMPPPPPPPEM          ]    DV     0]    ____________''xoooooO'''v",
+        "v....]mccm  <<  mPPm         ]++]]]]]]]]]]]   ,____________''xoooooO'''v",
+        "v....]           PP          ]           ;]    ____________''OOOOOOO'''v",
+        "v....]#####     <PP<         x            x   ,____________''''''''''''v",
+        "v....]<<A R     <PP<         x            x    ____________''OOOOOOO'''v",
+        "v....]    R      PP          x            x   ,____________''xoooooO'''v",
+        "v....]l   #     mPPm         x            x  b ____________''xoooooO'''vvvvvvvvvvvvvvvvvvvvvvvvv",
```
Comment (anothersimulacrum) [inline data/json/mapgen/steel_mill/steel_mill_z1.json]: No fence here, is that intentional?

Comment (Kanexan) [reply data/json/mapgen/steel_mill/steel_mill_z1.json]: As far as I can see, all of those lines have fences? Regardless, I'm in the process of updating the map; I'll be editing the fencing a bit, so I'll try to catch this. Thanks for pointing it out!

**Comment by Night-Pryanik** [comment]
Some screenshots would be really appreciated.

**Comment by Kanexan** [comment]
I'll add screenshots shortly, once I fix a couple of my mistakes with the overmap.

**Comment by Kanexan** [comment]
Added the screenshots; just tell me if you need more!

**Comment by curstwist** [comment]
Quick one. 1st screenshot. Should those utility shelves in offices be desks?

**Comment by Kanexan** [comment]
Oh crap, yes. I saw that and I fixed it in the file, but apparently I didn't actually get a screenshot of how they were supposed to be.

**Comment by Kanexan** [comment]
Alright, fixed the screenshot. Thanks for pointing that out!

File: data/json/mapgen_palettes/steel_mill_palette.json [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      "B": [ "t_buffer_stop" ],
+      ":": [ "t_open_air" ],
+      ";": [ "t_gates_control_concrete" ],
+      "a": [ "t_open_air_rooved" ],
+      "r": [ "t_metal_flat_roof" ],
+      "g": [ "t_generator_broken" ],
+      "G": [ "t_grate" ],
+      "2": [ "t_sai_box" ],
+      "5": [ "t_switchgear_l" ],
+      "`": [ "t_thconc_y"]
```
Comment (ZhilkinSerg) [inline data/json/mapgen_palettes/steel_mill_palette.json]:
```suggestion
      "`": [ "t_thconc_y" ]
```

File: data/json/mapgen_palettes/steel_mill_palette.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    },
+    "furniture": {
+      "h": [ "f_chair" ],
+      "T": [ "f_table" ],
+      "A": [ "f_air_conditioner" ],
+      "d": [ "f_desk" ],
+      "s": [ "f_utility_shelf"],
```
Comment (ZhilkinSerg) [inline data/json/mapgen_palettes/steel_mill_palette.json]:
```suggestion
      "s": [ "f_utility_shelf" ],
```

File: data/json/mapgen_palettes/steel_mill_palette.json [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      "C": [ "f_counter" ],
+      "W": [ "f_water_heater" ],
+      "w": [ "f_water_purifier" ],
+      "U": [ "f_home_furnace" ],
+      "f": [ "f_fridge" ],
+      "D": [ "f_vending_c" ],
+      "V": [ "f_vending_c" ],
+      "7": [ "f_filing_cabinet" ]
+    },
+    "toilets": { "t": { } },
```
Comment (ZhilkinSerg) [inline data/json/mapgen_palettes/steel_mill_palette.json]:
```suggestion
    "toilets": { "t": {  } },
```

**Comment by snipercup** [comment]
I added items and monsters. If anyone is working on this:

> In the future, I want to make the roof area more detailed, add the ~~monster~~, vehicle, ~~and item~~ spawn lists, and possibly create some additional heavy-industrial zombie types for a more detailed and realistic feel to the area.

So if anyone wants to work on this, add roof decoration and vehicles and maybe special monster, good luck!

**Comment by Kanexan** [comment]
Thank you very much. I am planning on returning to this, with both the unique zombie types and some additional spice to it, but due to personal reasons I haven't been able to complete the location. I had actually been working on a PR to remove the steel mill from spawn temporarily, as I knew that as-is it was a disappointment

**Pull request merged**